### ISSUE TITLE
Json: improve projection deduplication for scenarios containing collection indexers

### ISSUE THREAD

**Issue by maumar** [dotnet/efcore#29513]
Currently we de-duplicate projections only if the additional path doesn't contain any collection indexes. They can only appear in the "source" JSON path. We could potentially improve that. e.g.:

source: x.OwnedReference
dupe: x.OwnedReference.OwnedCollection[0]

We could also de-duplicate if source projects entire collection and the dupe projects just one of it's elements, e.g.:

source: x.OwnedCollection
dupe: x.OwnedCollection[2]

**Comment by smitpatel** [comment]
This should just work.
The dupe should have path segments `OwnedCollection` and `SqlConstant(2)`. Current logic accounts for it already.

**Comment by maumar** [comment]
to answer @roji 's question from the other thread:

> is this deduplication something we do in non-JSON cases (e.g. for owned collections or collection includes)?
For example:

> One = x.MyNonJsonCollection
> Two = x.MyNonJsonCollection.First()

Owned collections and non-owned entities have their own keys, whereas JSON collections use their position in the collection as part of the key. If the entities are no tracking we can just populate those keys with dummy values, but for tracking queries we would need the correct values, otherwise we will get incorrect results (collections added in wrong places etc).

On top of that, we can't use ChangeTracker to deduplicate in Tracking scenarios (even though the entity materializer code already does that for us) and leave no tracking with dummy values, because then we would have drastically different queries between tracking and no tracking. 

So there needs to be a mechanism that works in a similar way for tracking and no tracking, essentially we need to propagate information about indexers from JsonQueryExpression into shaper and as we generate the shaper code keep track entities/collections we already encountered (or rather variables we generated and assigned those entities/collections to)
